src/Interpreters/DuplicateDistinct.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <Functions/FunctionFactory.h>
+#include <IO/WriteHelpers.h>
+#include <Interpreters/InDepthNodeVisitor.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ASTSelectQuery.h>
+#include <Parsers/ASTSetQuery.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+#include <Parsers/IAST.h>
+#include <Common/typeid_cast.h>
+
+namespace DB
+{
+
+/// Removes duplicate DISTINCT from queries.
+class DuplicateDistinctMatcher
+{
+public:
+    struct Data
+    {
+        bool is_distinct;
+        std::vector<String> last_ids;
+    };
+
+    static void visit(const ASTPtr & ast, Data & data)
+    {
+        if (!ast)
+            return;
+
+        auto * select_query = ast->as<ASTSelectQuery>();
+        if (!select_query)
```

Comment:
It's better to place here only ast specific logic. All calls for casted type (ASTSelectQuery) should be inside it's `visit(ASTSelectQuery & )` function.